### PR TITLE
feat: log metadata/input/output/request/response

### PR DESCRIPTION
**Pull request by trivikr**
*Issue #, if available:*
Internal JS-2037

*Description of changes:*
log metadata/input/output/request/response

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by trivikr** [comment]
Build fails because of StackOverflow error discussed in https://github.com/aws/aws-sdk-js-v3/issues/1460

Confirmed that the [rebuild](https://us-west-2.console.aws.amazon.com/codesuite/codebuild/119327258862/projects/sdk-staging-test/build/sdk-staging-test%3Aad1a8669-3980-4eb4-bb93-01a3a56a4d59?#) is successful

File: packages/middleware-logger/src/loggerMiddleware.ts [hunk 1]
```diff
@@ -26,8 +55,9 @@ export const loggerMiddlewareOptions: FinalizeRequestHandlerOptions & AbsoluteLo
   step: "finalizeRequest",
 };
 
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (alexforsyth) [inline packages/middleware-logger/src/loggerMiddleware.ts:58]: Why do we need this?

Comment (alexforsyth) [reply packages/middleware-logger/src/loggerMiddleware.ts:58]: like why not just remove options?

Comment (trivikr) [reply packages/middleware-logger/src/loggerMiddleware.ts:58]: Was explained in commit description b5c87627123d0416bc5e8f7f51192635dac9a962

CodeGen passes options to getXPlugin commands, but loggerMiddleware reads logger from context. A long-term fix would need discussion on whether `getXPlugin` should work without config passed to it.

Comment (AllanFly120) [reply packages/middleware-logger/src/loggerMiddleware.ts:58]: Can we remove `options` parameter? I think it should not break the TS.

Comment (trivikr) [reply packages/middleware-logger/src/loggerMiddleware.ts:58]: It does break TS with:
```console
Expected 1 arguments, but got 0
```

**Pull request merged**